### PR TITLE
Set properties to null when none are available

### PR DESCRIPTION
**Pull request by CoderDake**
Addresses https://github.com/flutter/devtools/issues/5933

Certain multi view widgets don't have layout properties so the layout explorer wasn't updating properly.
This PR sets the properties to null when none are available and add's some state to communicate that to the layout explorer.

# Before Fix

https://github.com/flutter/devtools/assets/1386322/69cc62a1-61d9-4d5d-a118-a2e1eef162ca

# After Fix

https://github.com/flutter/devtools/assets/1386322/68b76db1-794c-4237-a4ce-3b0b1613ee53

### CODE REVIEW

**Comment by kenzieschmoll** [comment]
@jacob314 if I remember correctly, if a widget doesn't have any layout properties, is the layout explorer supposed to show the layout properties for its closest ancestor?

File: packages/devtools_app/lib/src/screens/inspector/layout_explorer/ui/layout_explorer_widget.dart [hunk 1]
```diff
@@ -80,6 +80,8 @@ abstract class LayoutExplorerWidgetState<W extends LayoutExplorerWidget,
       final newSelection = await fetchLayoutProperties();
       if (newSelection != null) {
         _setProperties(newSelection);
+      } else {
+        _setProperties(null);
       }
```
Comment (kenzieschmoll) [inline packages/devtools_app/lib/src/screens/inspector/layout_explorer/ui/layout_explorer_widget.dart:85]: this can be condensed to:
```dart
_setProperties(newSelection);
```

File: packages/devtools_app/release_notes/NEXT_RELEASE_NOTES.md [hunk 2]
```diff
@@ -9,7 +9,7 @@ Dart & Flutter DevTools - A Suite of Performance Tools for Dart and Flutter
 - Fix overlay notifications so they cover the area that their background blocks - [#5975](https://github.com/flutter/devtools/pull/5975)
 
 ## Inspector updates
-TODO: Remove this section if there are not any general updates.
+- Fix bug related to upcoming Flutter Multi View functionality - [#5986](https://github.com/flutter/devtools/pull/5986)
```
Comment (kenzieschmoll) [inline packages/devtools_app/release_notes/NEXT_RELEASE_NOTES.md:12]: I would argue that we don't need a release note for a bug that doesn't exist yet 🤔

Comment (CoderDake) [reply packages/devtools_app/release_notes/NEXT_RELEASE_NOTES.md:12]: I was split on this, thanks for your input! :D

**Comment by CoderDake** [comment]
@kenzieschmoll  
> if I remember correctly, if a widget doesn't have any layout properties, is the layout explorer supposed to show the layout properties for its closest ancestor?

This sound familiar based on some code that I stumbled upon, it was causing some weird behaviour since all of the parents here don't have layout data.

If it turns out that null properties is an expected behaviour then I could look at this a bit deeper to see if I could be more specific to only show the message if no parents have layout info either.

**Comment by jacob314** [comment]
> @jacob314 if I remember correctly, if a widget doesn't have any layout properties, is the layout explorer supposed to show the layout properties for its closest ancestor?

That is correct. In general I thought that happened automatically as each Element got an associated RenderObject even if that was just the one shared with its ancestor.

**Comment by jacob314** [comment]
> 

Perhaps this changed with @goderbauer's refactors to support multiple windows. We should rethink what makes sense. Generally what we are looking for is just the RenderObject that best approximates the space the widget takes up. Potentially now we need to go down the tree instead of up but would be good to have a principled design in what we are trying to do. Some special cases like Root might no longer have a real size given multiple windows and that is fine.

**Comment by goderbauer** [comment]
With the multi view changes it makes sense that some widgets do not have layout properties anymore. Prior to this change, every element always had an associated render object (i.e. `Element. findRenderObject()` would always find something). Since RenderObjects define the layout, you could say that each element had layout properties.

In the multi view world some parts of the widget tree are no longer backed by render objects. This is due to the fact that we have a single unified widget tree describing the content for all views. To this widget tree, multiple render trees (one for each view) are attached, but some widgets/elements may not have a render object at all (and `Element. findRenderObject()` will return null for those). Here is an example widget tree to illustrate:

```dart
MyDataProvider(
  data: _myAppData,
  child: ViewCollection(
    views: <Widget>[
      View(
        view: <...>
        child: ColoredBox(color: Colors.green),
      ),
      View(
        view: <...>
        child: ColoredBox(color: Colors.blue),
      ),
    ],
  )
);
```

This is a single widget tree defining the content of two Views (one green, one blue). The View widgets each bootstrap their own render trees, so if we were to call `debugDumpRenderTree` on this app, we would see two render trees looking something like this:

```
// render tree #1
RenderView#1
  _RenderColoredBox(color: green)

// render tree #2
RenderView#2
  _RenderColoredBox(color: blue)
```

The widgets above the `View` widgets (here: `MyDataProvider` and `ViewCollection`) don't define any layout aspects of the app, they are just there to inject some data (`MyDataProvider`) and to give our views anchor points in the widget tree (`ViewCollection`). You also can't associate them with a single render object (would you chose `RenderView#1` or `#2)? Neither choice makes sense here because while they organize those RenderObjects, they don't take on their layout properties. In other words, `ViewCollection` and `MyDataProvider` don't occupy any space in the app's layout.

tl;dr: It makes sense to me to disable the layout explorer for widgets that are not backed by a render tree (i.e. widgets that appear above the start of the render tree). The signal I would use is does `Element.findRenderObject` return null or not?

**Comment by goderbauer** [comment]
Note, though, that I would expect widgets that are clearly part of a render tree to still show the layout explorer. For example, if the `MyDataProvider` from the previous example were to be used inside a View...

```dart
View(
  view: <..>
  child: MyDataProvider(
    data: <..>
    child: SizedBox(),
  ),
);
```

... I would expect that the layout explorer shows me the layout properties of the `SizedBox` render object even when I select `MyDataProvider` (that's also the renderobject that `findRenderObject` called on MyDataProvider's element would return). Here, `MyDataProvider` is part of the layout of this app.

File: packages/devtools_app/lib/src/screens/inspector/layout_explorer/ui/layout_explorer_widget.dart [hunk 3]
```diff
@@ -80,6 +80,8 @@ abstract class LayoutExplorerWidgetState<W extends LayoutExplorerWidget,
       final newSelection = await fetchLayoutProperties();
```
[the remark this answers is not in the document]
Comment (CoderDake) [reply packages/devtools_app/lib/src/screens/inspector/layout_explorer/ui/layout_explorer_widget.dart:80]: It looks like `fetchLayoutProperties` calls `WidgetInspectorServiceExtensions.getLayoutExplorerNode` is called through objectGroups.
https://github.com/flutter/devtools/blob/54111ea25e1067a1735bac07b888f90d85a7df37/packages/devtools_app/lib/src/screens/inspector/layout_explorer/ui/layout_explorer_widget.dart#L107

[the remark this answers is not in the document]
Comment (jacob314) [reply packages/devtools_app/lib/src/screens/inspector/layout_explorer/ui/layout_explorer_widget.dart:80]: Here is the relevant code on the Flutter side. It is using the Element.renderObject getter.
          final RenderObject? renderObject = value is Element ? value.renderObject : null;
https://github.com/flutter/flutter/blob/c8265f81cf591587f6056a9f9def6150008d9b6e/packages/flutter/lib/src/widgets/widget_inspector.dart#L2070C64-L2070C64
and then adding some properties if the render object has been laid out.

https://github.com/flutter/flutter/blob/c8265f81cf591587f6056a9f9def6150008d9b6e/packages/flutter/lib/src/widgets/widget_inspector.dart#L2102
We're just

[the remark this answers is not in the document]
Comment (goderbauer) [reply packages/devtools_app/lib/src/screens/inspector/layout_explorer/ui/layout_explorer_widget.dart:80]: Makes sense, findRenderObject is implemented in terms of the renderObject getter (with some additional checks). so this would behave as expected then. If this gives you null we shouldn't show the layout explorer.

**Comment by CoderDake** [comment]
> Note, though, that I would expect widgets that are clearly part of a render tree to still show the layout explorer. For example, if the `MyDataProvider` from the previous example were to be used inside a View...
> 
> ```dart
> View(
>   view: <..>
>   child: MyDataProvider(
>     data: <..>
>     child: SizedBox(),
>   ),
> );
> ```
> 
> ... I would expect that the layout explorer shows me the layout properties of the `SizedBox` render object even when I select `MyDataProvider` (that's also the renderobject that `findRenderObject` called on MyDataProvider's element would return). Here, `MyDataProvider` is part of the layout of this app.

@jacob314 I've recreated goderbauer's provider example here and noticed that we The parent of the builder is the context that we see, not the child.

I've grabbed a screenshot to show what it looks like, does this look like it is working like you would expect it?

Screenshot 2023-07-13 at 5 11 37 PM

**Comment by goderbauer** [comment]
I think this matches my expectation, yes. (I think in my original comment I had up and down mixed up, showing the parent is the right thing, though)

**Comment by jacob314** [comment]
If you select a widget thats parent is a Row/Column we show that parents and its children. What we try to do is highlight the child that you selected but we clearly aren't doing a good job in that case. Seems like the vertical scroll in the layout explorer is off so you can't see that the child it is scrolled to is the ValueListenableBuilder.
What we are trying to do is explain why the ValueListenableBuilder is laid out the way it is and to do that we need to show how it is arranged in the Column.

**Review by jacob314 — APPROVED** [review]
lgtm
